src/button/common.c

```diff
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: GPL-2.0-only
+#include <stdio.h>
+#include "button/common.h"
+#include "common/dir.h"
+#include "config/rcxml.h"
+
+char *
+button_filename(const char *name)
+{
+	static char buf[4096] = { 0 };
+	snprintf(buf, sizeof(buf), "%s/%s", theme_dir(rc.theme_name), name);
+	return buf;
+}
```

Comment:
I am not sure about returning statically allocated memory here. If we want to keep doing that it should be made very clear in the header doc that this is the case (e.g. `xstrdup()` when wanting to store / never `free()`).

Nitpick: the header calls the argument `filename` which doesn't match the implementation.

Will look into the PR in more detail tomorrow.

Reply:
I have `s/filename/name`. Good catch.

I don't see the problem with returning a static variable. I've moved it to translation unit level - but I think either way is fine.

It's like doing `return "foo"`. The calling function can use it but must not try to free() it.

I'm more worried about any function that allocates memory on the heap - because the caller much free it.

We have a few more of these in other places. If memory was super tight I guess we could declare a buf[4096] in the calling function and just pass the pointer to `button_filename()`

Reply:
My issue is more about not necessarily knowing what happens when calling some function that returns a `char *`, is it returning static memory or heap memory? Maybe we should go over all the public function that return either and specify it in the header files.

Returning statically allocated memory here seems fine, for me its really more about the question that come with it when using:
- "Is it leaking memory if I don't free it?"
- "Can is store a pointer to this somewhere or will the content get overwritten by future invocations?"

For me this then usually means that I usually look into the implementation to answer these questions but when somebody using a function like that the first time, that person might expect a specific behavior which may not be the case.

Reply:
This is a really bad practice. Somewhere in the future we will call this function twice and overwrite the previous result.

Reply:
I don't actually think its bad practice at all. Callers just need to know if they have to `strdup()` themselves or not. Same for `free()` when not using statically allocated memory.

Reply:
We had this issue with multiple functions in libc over the last 20 years. This is really, really, really bad idea.

Reply:
Me guilty here. Will fix. Will probably put buf at highest frame. Need to have dinner first though :smile:

Reply:
`static char buf[4096] = { 0 };` fixed.